### PR TITLE
tests: rewrite in go and simplify TLSChallengeTest

### PR DESCRIPTION
**Pull request by porridge**
### Description

This supersedes https://github.com/stackrox/stackrox/pull/11986 which accumulated hundreds of retry comments making it hard to follow the review thread.

`TLSChallengeTest` is currently by far the most flaky test owned by draco team. Rather than invest more time into debugging the groovy test and the k8s interface library (where the problem seems to lie) I decided to kill two birds with one stone and rewrite this test in Go. This starts to pave the road towards creating an improved e2e testing framework for nongroovy.

While rewriting I also reused the additional CA already in use for `TestCASetup` which speeds up and simplifies the test considerably.

This PR resolves:
- https://issues.redhat.com/browse/ROX-24631(Move TLSChallengeTest to the nongroovy job)
- https://issues.redhat.com/browse/ROX-22594 (TLSChallengeTest / Verify sensor can communicate with central behind an untrusted load balancer FAILED)
And partially addresses:
- https://issues.redhat.com/browse/ROX-9232 (Create end2end testing framework for Go)

### User-facing documentation

(*must be* 2 items and both *must be* checked)

- [x] CHANGELOG update is not needed
- [x] Documentation is not needed

### Testing

- [x] inspected CI results

#### Automated testing

(*must be* at least 1 item and all items *must be* checked)

- [x] added e2e tests
- [x] modified existing tests
  

#### How I validated my change

CI should be enough. I had the bot re-run this test a few times and saw a single related flake that I'm addressing in https://github.com/stackrox/stackrox/pull/12305. The other failures seem unrelated (collector crash, `Test_ComplianceOperatorScanConfigSync` failure, and a stackrox setup failure on GKE).

### CODE REVIEW

File: tests/common.go [hunk 1]
```diff
@@ -43,6 +56,33 @@ var (
 	log = logging.LoggerForModule()
 )
 
+// logf logs using the testing logger, prefixing a high-resolution timestamp.
```
Comment (janisz) [inline tests/common.go:59]: New functions are used only in TLS test do we need them in common?

File: tests/common_test.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
+//go:build test_e2e
```
[the remark this answers is not in the document]
Comment (porridge) [reply tests/common_test.go:1]: This won't work:
```
+ roxvet
# github.com/stackrox/rox/tests
# [github.com/stackrox/rox/tests]
roxvet: tests/common_test.go:20:20: undefined: logMatcher
make: *** [Makefile:782: roxvet] Error 123
```

**Comment by porridge** [comment]
Comments addressed, PTAL @janisz

File: tests/common_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (janisz) [reply tests/common_test.go:1]: then common.test also can be without buld flag. Why do we need to guard it if it does not contain any e2e tests?

File: tests/tls_challenge_test.go [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+	appsV1 "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+	apiErrors "k8s.io/apimachinery/pkg/api/errors"
+	metaV1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+const (
+	s                  = namespaces.StackRox // for brevity
+	proxyNs            = "qa-tls-challenge"  // Must match the additionalCA X509v3 Subject Alternative Name
```
Comment (janisz) [inline tests/tls_challenge_test.go:28]: In multiple functions we shadow this constant. Why? If it's a constant why not just use it?

File: tests/tls_challenge_test.go [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+	nginxLabels := map[string]string{"app": "nginx"}
+	nginxTLSSecretName := "nginx-tls-conf"
+	nginxConfigName := "nginx-proxy-conf"
+	ts.logf("Setting up nginx proxy in namespace %q...", proxyNs)
+	ts.createProxyNamespace(proxyNs)
+	ts.installImagePullSecret(proxyNs)
+	ts.createProxyTLSSecret(proxyNs, nginxTLSSecretName)
+	ts.createProxyConfigMap(proxyNs, centralEndpoint, nginxConfigName)
+	ts.createService(ts.ctx, proxyNs, name, nginxLabels, map[int32]int32{443: 8443})
+	ts.createProxyDeployment(proxyNs, name, nginxLabels, nginxConfigName, nginxTLSSecretName)
```
Comment (janisz) [inline tests/tls_challenge_test.go:109]:
```suggestion
	ts.createProxyNamespace()
	ts.installImagePullSecret()
	ts.createProxyTLSSecret(nginxTLSSecretName)
	ts.createProxyConfigMap(centralEndpoint, nginxConfigName)
	ts.createService(ts.ctx, name, nginxLabels, map[int32]int32{443: 8443})
	ts.createProxyDeployment(name, nginxLabels, nginxConfigName, nginxTLSSecretName)
```

File: tests/tls_challenge_test.go [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+		return
+	}
+	ts.Require().NoError(err, "cannot create proxy namespace %q", proxyNs)
+}
+
+func (ts *TLSChallengeSuite) installImagePullSecret(proxyNs string) {
+	configBytes, err := json.Marshal(config.DockerConfigJSON{
+		Auths: map[string]config.DockerConfigEntry{
+			"https://quay.io": {
+				Username: os.Getenv("REGISTRY_USERNAME"),
```
Comment (janisz) [inline tests/tls_challenge_test.go:125]: Can this be empty?
https://github.com/stackrox/stackrox/blob/0704c5ec5473b5de50f3f9ffcbea2715fcfcb989/pkg/testutils/centralgrpc/connect_to_central.go#L39

**Review by janisz — APPROVED** [review]
LGTM, except shadowing const when it's not needed

File: tests/common_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (porridge) [reply tests/common_test.go:1]: I'm not sure I understand your question. Did you mean to ask why `common.go` is guarded? I don't know :-)
Let me try changing that in a followup PR to see if there is a reason.

File: tests/tls_challenge_test.go (hunk 3 above)
Comment (porridge) [reply tests/tls_challenge_test.go:28]: Good point, let me follow up on this in https://github.com/stackrox/stackrox/pull/12305 for the sake of progress.

File: tests/tls_challenge_test.go (hunk 5 above)
Comment (porridge) [reply tests/tls_challenge_test.go:125]: Right, it probably could be improved to abort early rather than produce a useless secret. Again, let me follow up on this in next PR.

File: tests/common_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (porridge) [reply tests/common_test.go:1]: [This](https://github.com/stackrox/stackrox/blob/797db99473d5c8004c4fb545f5aa9208f32c89d6/tools/roxvet/analyzers/testtags/analyzer.go#L14) is the reason. I'm planning to take a stab at changing the approach here but this is not a simple change.
